### PR TITLE
[Bugfix:TAGrading] Grade button and progress bar

### LINKED ISSUE #9846: red REGRADE button despite TA grading being 'complete'
**Describe the bug**

On the gradeables page for a specific homework, the REGRADE button is red, and the progress bar is incomplete, suggesting TA grading is incomplete and overdue:

Screenshot 2023-10-04 at 11 08 19 AM

Also, when viewing a specific student, there is an indication that grading is incomplete:
Screenshot 2023-10-04 at 11 24 08 AM

However on the stats page,

All on time submissions have been graded
Screenshot 2023-10-04 at 11 06 13 AM

And there are no unresolved grade inquiries
Screenshot 2023-10-04 at 11 06 22 AM

And the stats filter is not configured to include any corner case 
Screenshot 2023-10-04 at 11 06 39 AM

I know we've discussed having all of the ta grading pages using the settings from that stats filter.  What is the status of that implementation?

Also, perhaps this case is red because there are version conflicts?   Maybe we forgot to include that in the stats filter?

**Expected behavior**

If the button is red / progress meter incomplete... it should be clear WHY on the stats page.  

- [ ] add text with "# of version conflicts" to the stats page

- [ ] add "include version conflicts" to the stats filter (maybe this doesn't make sense?)

- [ ] on the gradeables page, make the button color & progress meters use the settings of the ta grading stats filter

- [ ] when on the ta grading page for a specific student, make the progress meter use the settings of the ta grading stats filter

(some of these items may duplicate other issues)

### PR DESCRIPTION
**Pull request by VVVITH**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?

Fixes #9846 
The grade button and the progress button for TAs is always using the data of all submissions.
In this test case:
截屏2023-11-10 13 44 26
The wrong behavior:
截屏2023-11-10 13 49 21

### What is the new behavior?
Now, the data used by grade button and progress button will affect by TA's personal setting as they choose in the Gradeable -> Grading Stats -> Edit fliters page.
The same test case:
截屏2023-11-10 13 44 26
Now:
截屏2023-11-10 13 45 50

### CODE REVIEW

File: site/app/models/gradeable/Gradeable.php [hunk 1]
```diff
@@ -1902,21 +1910,30 @@ public function getGradingProgress(User $grader) {
                 $sections[$key] = [
                     'total_components' => $value * $num_components,
                     'graded_components' => 0,
+                    'non_late_total_components' => ($value  - (array_key_exists($key, $late_submitted) ? $late_submitted[$key] : 0) ) * $num_components,
```
Comment (williamjallen) [inline site/app/models/gradeable/Gradeable.php:1913]: This is a great opportunity to use PHP's [null coalescing](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.coalesce) operator:
```suggestion
                    'non_late_total_components' => ($value  - ($late_submitted[$key] ?? 0) ) * $num_components,
```

File: site/app/templates/grading/electronic/ta_status/StatusBase.twig [hunk 2]
```diff
@@ -363,19 +363,27 @@
         var bad_submissions_status = '{{ include_bad_submissions }}';
         if (bad_submissions_status == 'omit') {
             document.cookie = "include_bad_submissions=include;";
+            const cookie_string = "include_bad_submissions__" + '{{ gradeable_id }}';
```
Comment (williamjallen) [inline site/app/templates/grading/electronic/ta_status/StatusBase.twig:366]: You don't need to use string concatenation here:
```suggestion
            const cookie_string = "include_bad_submissions__{{ gradeable_id }}";
```

Apply below as well.

File: site/app/views/NavigationView.php [hunk 3]
```diff
@@ -661,7 +661,14 @@ private function getGradeButton(Gradeable $gradeable, int $list_section) {
             }
 
             if (!$gradeable->hasDueDate()) {
-                $progress_bar = $gradeable->getGradingProgress($this->core->getUser());
+
+                $cookie_string = "include_bad_submissions__" . $gradeable->getId() ;
+                $bad_submissions = array_key_exists($cookie_string, $_COOKIE) ? ($_COOKIE[$cookie_string]==="include" ? true : false) : false;
+
+                $cookie_string = "include_null_section__" . $gradeable->getId() ;
+                $null_section = array_key_exists($cookie_string, $_COOKIE) ? ($_COOKIE[$cookie_string]==="include" ? true : false) : false;
```
Comment (williamjallen) [inline site/app/views/NavigationView.php:669]: Here's another good chance to clean up the code with the null coalescing operator:
```suggestion
                $null_section = ($_COOKIE[$cookie_string] ?? '') === "include";
```

File: site/app/views/NavigationView.php [hunk 4]
```diff
@@ -710,7 +717,14 @@ private function getGradeButton(Gradeable $gradeable, int $list_section) {
 
             if ($gradeable->getType() === GradeableType::ELECTRONIC_FILE) {
                 if ($gradeable->isTaGrading()) {
-                    $TA_percent = $gradeable->getGradingProgress($this->core->getUser());
+
+                    $cookie_string = "include_bad_submissions__" . $gradeable->getId();
+                    $bad_submissions = array_key_exists($cookie_string, $_COOKIE) ? ($_COOKIE[$cookie_string]==="include" ? true : false) : false;
+
+                    $cookie_string = "include_null_section__" . $gradeable->getId() ;
+                    $null_section = array_key_exists($cookie_string, $_COOKIE) ? ($_COOKIE[$cookie_string]==="include" ? true : false) : false;
```
Comment (williamjallen) [inline site/app/views/NavigationView.php:725]: Same here

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/views/NavigationView.php (hunk 4 above)
Comment (VVVITH) [reply site/app/views/NavigationView.php]: thanks for your suggestion, the code have changed

**Review by ANS1514 — CHANGES_REQUESTED** [review]
Aside from the linting errors, I saw correct behavior in the code.

**Review by ZheyuDeng — APPROVED** [review]
Tested the code. Worked as described!

**Review by benjaminlee755 — APPROVED** [review]
Code looks good and everything works like it is supposed to.

**Review by williamjallen — APPROVED** [review]
Code looks good. I did not review the functionality.

**Review by ANS1514 — APPROVED** [review]

**Pull request merged**